File: bend2/effs/tcp_recv_buf.c
// TCP
// ===

// TCP.recv_bytes into a Bytes: the bytes received are copied once into
// one packed block (io_buf), a byte a cell, instead of a List cell each.
static Term tcp_recv_buf_pack(Env e, IoWork* w) {
  Term r = w->code ? io_fail(e, w->code, NULL)
    : io_done(e, io_buf(e, w->data, w->size));
  free(w->data);
  return io_tup(e, io_hand(w->hand), r);
}

// The loop parked the request until the socket was readable; a recv
// that still finds nothing (the socket is non-blocking) parks again.
static Term tcp_recv_buf_more(Env e, IoWork* w) {
  int   fd  = (int)w->hand;
  short dir = POLLIN;
  w->size = io_sys_end(w, io_wire_read(fd, w->data, (size_t)w->made, &dir));
  return w->code == EAGAIN ? io_wait_on(w, fd, dir, 0, tcp_recv_buf_more)
    : tcp_recv_buf_pack(e, w);
}

Term tcp_recv_buf_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  w->made = f[1] < INT32_MAX ? (intptr_t)f[1] : INT32_MAX;
  w->data = io_mem(malloc((size_t)w->made + 1));
  return tcp_recv_buf_more(e, w);
}

static void __attribute__((constructor)) tcp_recv_buf_use(void) {
  io_eff(CID_TCP_RECV_BUF, tcp_recv_buf_run, IO_READ);
}

File: bend2/effs/tcp_recv_buf.js
// TCP
// ===

// TCP.recv_bytes into a Bytes: a string of char codes 0..255, the bytes
// as they are, no text decoder anywhere near the socket.
function tcp_recv_buf(socket, max, k) {
  const sys = io_sys();
  const fd = socket;
  const b = new Uint8Array(Math.max(Number(max), 1));
  const again = sys.mac ? 35 : 11;
  const go = () => {
    const n = Number(sys.recv(fd, sys.ptr(b), Number(max), 0));
    if (n < 0) {
      const code = sys.errno();
      if (code === again) {
        io_park_on(fd, false, k, go);
        return undefined;
      }
      return io_tup(socket, io_fail(code));
    }
    let s = "";
    for (let i = 0; i < n; i += 8192) {
      s += String.fromCharCode.apply(null, b.subarray(i, Math.min(n, i + 8192)));
    }
    return io_tup(socket, io_done(s));
  };
  return go();
}

function tcp_recv_buf_need() {
  return { read: true };
}
